css/css-images/radial-gradient-transition-hint-crash.html WPT asserts on debug builds

obyknovenius opened the pull request:
#### 1ca6deb96ab8429e9ce48f9576801100210893c7
<pre>
css/css-images/radial-gradient-transition-hint-crash.html WPT asserts on debug builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=277218">https://bugs.webkit.org/show_bug.cgi?id=277218</a>

Reviewed by Sam Weinig.

`StyleGradientImage::computeStops()`: Don&apos;t re-compute midpoint offsets.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::StyleGradientImage::computeStops const):

Canonical link: <a href="https://commits.webkit.org/284231@main">https://commits.webkit.org/284231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/463c5861369eaa2d85c563442cfcad35e0b3ea33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72842 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19917 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54812 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13251 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44006 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59380 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35279 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16806 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18275 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62620 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17154 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74536 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16399 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62299 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59462 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62335 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15272 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10296 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3906 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43965 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45039 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46233 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44781 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/rendering/style/StyleGradientImage.cpp:
```diff
@@ -945,11 +945,15 @@ template<typename GradientAdapter, typename Stops> GradientColorStops StyleGradi
+        return { *stop.offset, stop.color };
+    });
+
+    std::ranges::stable_sort(result, [](const auto& a, const auto& b) {
```
Explicitly sorting seems reasonable, but the assumption has been they were already sorted. Do you know what causes the implicit sort to fail?

obyknovenius replied:
I've finally figured it out. The actual issue is on the line https://github.com/WebKit/WebKit/blob/525fe6257f8625fbb17f24c28a06ff232490b76d/Source/WebCore/rendering/style/StyleGradientImage.cpp#L807
While the formula is correct, dealing with floats can introduce imprecision problems.
If you add a debug line like this:
```cpp
fprintf(stderr, "newStops[%zu].offset = offset1 + (offset - offset1) * (7 + y) / 13 = %f + %f * %lu = %f + %f = %f\n", y, offset1, (offset - offset1), (7 + y) / 13, offset1, (offset - offset1) * (7 + y) / 13, *newStops[y].offset);
```
you'll see the issue:
```
newStops[6].offset = offset1 + (offset - offset1) * (7 + y) / 13 = -15403597.000000 + 15403597.000000 * 1 = -15403597.000000 + 15403598.000000 = 1.000000
```
Luckily, we don't have to recompute the offset of midpoins. We can simply reuse the original value. I'll update the PR accordingly.

weinig reviewed (approved):
Nice!

obyknovenius committed:
css/css-images/radial-gradient-transition-hint-crash.html WPT asserts on debug builds
https://bugs.webkit.org/show_bug.cgi?id=277218

Reviewed by Sam Weinig.

`StyleGradientImage::computeStops()`: Don't re-compute midpoint offsets.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::StyleGradientImage::computeStops const):

Canonical link: https://commits.webkit.org/284231@main
